File: applicatie/src/main/java/com/vdzon/administratie/dto/BoekingDto.java
package com.vdzon.administratie.dto;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.vdzon.administratie.model.boekingen.Boeking;
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetAfschrift;
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetDeclaratie;
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetFactuur;
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetRekening;
import lombok.*;

@ToString
@EqualsAndHashCode
@Getter
@Builder(toBuilder = true)
@AllArgsConstructor
@NoArgsConstructor

@JsonIgnoreProperties
public class BoekingDto {

    private String uuid;
    private String omschrijving;
    private String factuurNummer;
    private String rekeningNummer;
    private String afschriftNummer;
    private String declaratieNummer;

    public BoekingDto(Boeking boeking) {
        this.omschrijving = boeking.getOmschrijving();
        this.uuid= boeking.getUuid();
        if (boeking instanceof BoekingMetAfschrift){
            this.afschriftNummer = ((BoekingMetAfschrift) boeking).getAfschriftNummer();
        }
        if (boeking instanceof BoekingMetFactuur){
            this.factuurNummer = ((BoekingMetFactuur) boeking).getFactuurNummer();
        }
        if (boeking instanceof BoekingMetRekening){
            this.rekeningNummer = ((BoekingMetRekening) boeking).getRekeningNummer();
        }
        if (boeking instanceof BoekingMetDeclaratie){
            this.declaratieNummer = ((BoekingMetDeclaratie) boeking).getDeclaratieNummer();
        }
    }

}

File: model/src/main/java/com/vdzon/administratie/dto/BoekingDto.kt
package com.vdzon.administratie.dto

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import com.vdzon.administratie.model.boekingen.Boeking
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetAfschrift
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetFactuur
import com.vdzon.administratie.model.boekingen.relaties.BoekingMetRekening

@JsonIgnoreProperties
class BoekingDto(
        val uuid: String = "",
        val omschrijving: String = "",
        val factuurNummer: String? = null,
        val rekeningNummer: String? = null,
        val afschriftNummer: String? = null,
        val declaratieNummer: String? = null) {

    companion object {
        fun toDto(boeking: Boeking) = BoekingDto(
                omschrijving = boeking.omschrijving,
                uuid = boeking.uuid,
                afschriftNummer = if (boeking is BoekingMetAfschrift) boeking.afschriftNummer else null,
                factuurNummer = if (boeking is BoekingMetFactuur) boeking.factuurNummer else null,
                rekeningNummer = if (boeking is BoekingMetRekening) boeking.rekeningNummer else null
        )
    }

}
